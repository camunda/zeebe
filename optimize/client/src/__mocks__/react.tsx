/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

// @ts-ignore
export * from 'react';
export {default} from 'react';

const outstandingEffects: (() => void)[] = [];
const outstandingCleanups: (() => void)[] = [];

export const useEffect = (fn: () => (() => void) | void) => outstandingEffects.push(fn);
export const runLastEffect = () => {
  if (outstandingEffects.length) {
    const cleanup = outstandingEffects.pop()?.();

    if (cleanup) {
      outstandingCleanups.push(cleanup);
    }
  }
};

export const runLastCleanup = () => {
  if (outstandingCleanups.length) {
    outstandingCleanups.pop()?.();
  }
};

export const runAllEffects = () => {
  const numberOfEffects = outstandingEffects.length;

  for (let i = 0; i < numberOfEffects; i++) {
    const cleanup = outstandingEffects.shift()?.();

    if (cleanup) {
      outstandingCleanups.push(cleanup);
    }
  }
};

beforeEach(() => {
  outstandingEffects.length = 0;
});

export const runAllCleanups = () => {
  const numberOfEffects = outstandingCleanups.length;

  for (let i = 0; i < numberOfEffects; i++) {
    outstandingCleanups.shift()?.();
  }
};
