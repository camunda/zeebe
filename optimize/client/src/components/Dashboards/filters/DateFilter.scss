.DateFilter__Dashboard {
  width: 190px;
  &.fixed {
    width: max-content;
  }

  .filterMenu {
    width: 100%;
  }

  .to {
    margin-inline: 2px;
  }

  .Popover .popoverContent {
    width: auto;
    max-width: unset;
  }

  .DateFields {
    .PickerDateInput {
      &::after {
        display: none;
      }
    }
    .dateRangeContainer {
      position: relative;
      box-shadow: none;
      border: none;
      top: 0;
      background-color: initial;

      .rdrCalendarWrapper {
        background-color: initial;
      }

      &::before,
      &::after {
        display: none;
      }
    }
  }
}
