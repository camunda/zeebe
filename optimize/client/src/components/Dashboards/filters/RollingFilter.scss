.RollingFilter {
  min-width: 360px;

  .number,
  .Select {
    width: 100%;
  }
}
