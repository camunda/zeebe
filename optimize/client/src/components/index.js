/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

export {Analysis} from './Analysis';
export {Footer} from './Footer';
export {Header} from './Header';
export {PrivateRoute} from './PrivateRoute';
export {Dashboard} from './Dashboards';
export {Report} from './Reports';
export {Sharing} from './Sharing';
export {Home, Collection} from './Home';
export {default as Logout} from './Logout';
export {Processes, ProcessReport} from './Processes';
