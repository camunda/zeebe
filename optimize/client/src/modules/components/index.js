/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

export {Table} from './Table';
export {Button} from './Button';
export {Modal} from './Modal';
export {Dropdown} from './Dropdown';
export {Deleter} from './Deleter';
export {Input} from './Input';
export {Select} from './Select';
export {BPMNDiagram, ClickBehavior, PartHighlight} from './BPMNDiagram';
export {DMNDiagram} from './DMNDiagram';
export {CopyToClipboard} from './CopyToClipboard';
export {ReportRenderer, NoDataNotice} from './ReportRenderer';
export {ErrorBoundary} from './ErrorBoundary';
export {Icon} from './Icon';
export {ActionItem} from './ActionItem';
export {ShareEntity} from './ShareEntity';
export {DashboardRenderer, DiagramScrollLock} from './DashboardRenderer';
export {DefinitionSelection, VersionPopover, TenantPopover} from './DefinitionSelection';
export {TargetValueBadge} from './TargetValueBadge';
export {ErrorPage} from './ErrorPage';
export {Labeled} from './Labeled';
export {LabeledInput} from './LabeledInput';
export {DatePicker} from './DatePicker';
export {LoadingIndicator} from './LoadingIndicator';
export {ColorPicker} from './ColorPicker';
export {EntityName} from './EntityName';
export {EntityNameForm} from './EntityNameForm';
export {SubNav, NavItem} from './navigation';
export {HeatmapOverlay} from './HeatmapOverlay';
export {EntityList} from './EntityList';
export {UserTypeahead} from './UserTypeahead';
export {DateRangeInput} from './DateRangeInput';
export {LastModifiedInfo} from './LastModifiedInfo';
export {ReportDetails} from './ReportDetails';
export {InstanceCount} from './InstanceCount';
export {DocsLink} from './DocsLink';
export {Tabs} from './Tabs';
export {Checklist} from './Checklist';
export {Tooltip} from './Tooltip';
export {SelectionPreview} from './SelectionPreview';
export {DownloadButton} from './DownloadButton';
export {BulkDeleter} from './BulkDeleter';
export {AlertModal} from './AlertModal';
export {AlertsDropdown} from './AlertsDropdown';
export {DurationChart} from './DurationChart';
export {ReportTemplateModal, DashboardTemplateModal} from './TemplateModal';
export {KpiCreationModal} from './KpiCreationModal';
export {EmptyState} from './EmptyState';
export {PageTitle} from './PageTitle';
export {TextEditor} from './TextEditor';
export {EntityDescription} from './EntityDescription';
export {Popover} from './Popover';
export {InstanceViewTable} from './InstanceViewTable';
export {TenantInfo} from './TenantInfo';
export {TargetSelection} from './TargetSelection';
export {Loading} from './Loading';
