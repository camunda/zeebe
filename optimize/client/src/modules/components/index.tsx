/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

export {Button} from './Button';
export {Tabs} from './Tabs';
export {Icon} from './Icon';
export {ActionItem} from './ActionItem';
export {Input} from './Input';
export {LabeledInput} from './LabeledInput';
export {Labeled} from './Labeled';
export {Tooltip} from './Tooltip';
export {Modal} from './Modal';
export {Dropdown} from './Dropdown';
export {Select} from './Select';
export {TextEditor} from './TextEditor';
export {PageTitle} from './PageTitle';
export {EntityDescription} from './EntityDescription';
export {EntityNameForm} from './EntityNameForm';
export {LoadingIndicator} from './LoadingIndicator';
export {DocsLink} from './DocsLink';
export {UserTypeahead} from './UserTypeahead';
export {BPMNDiagram, ClickBehavior} from './BPMNDiagram';
export {DMNDiagram} from './DMNDiagram';
export {DownloadButton} from './DownloadButton';
export {NoDataNotice} from './ReportRenderer';
export {Table} from './Table';
export {TargetValueBadge} from './TargetValueBadge';
export {DatePicker} from './DatePicker';
export {DateRangeInput} from './DateRangeInput';
export {Checklist} from './Checklist';
export {DurationChart} from './DurationChart';
export {ColorPicker} from './ColorPicker';
export {Popover} from './Popover';
export {TenantInfo} from './TenantInfo';
export {VersionPopover, TenantPopover} from './DefinitionSelection';
export {CopyToClipboard} from './CopyToClipboard';
export {ShareEntity} from './ShareEntity';
export {ErrorPage} from './ErrorPage';
export {ErrorBoundary} from './ErrorBoundary';
export {TargetSelection} from './TargetSelection';
// @ts-ignore
export {DefinitionSelection} from './DefinitionSelection';
// @ts-ignore
export {ReportRenderer} from './ReportRenderer';
export {Loading} from './Loading';
export {EntityList} from './EntityList';
export {EmptyState} from './EmptyState';

export type {InputProps} from './Input';
export type {LabeledProps} from './Labeled';
export type {User} from './UserTypeahead';
export type {Canvas, ModdleElement, RegistryElement} from './BPMNDiagram';
export type {TooltipProps} from './Tooltip';
export type {TableBody, TableHead} from './Table';
