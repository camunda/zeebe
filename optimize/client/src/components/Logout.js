/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {useEffect} from 'react';
import {withRouter} from 'react-router-dom';
import {get} from 'request';
import {withErrorHandling} from 'HOC';
import {addNotification} from 'notifications';
import {t} from 'translation';

export function Logout({mightFail, history}) {
  useEffect(() => {
    (async () => {
      await mightFail(
        get('api/authentication/logout'),
        () => addNotification({text: t('navigation.logoutSuccess')}),
        () => addNotification({text: t('navigation.logoutFailed'), type: 'error'})
      );
      history.replace('/');
    })();
  }, [mightFail, history]);

  return null;
}

export default withRouter(withErrorHandling(Logout));
