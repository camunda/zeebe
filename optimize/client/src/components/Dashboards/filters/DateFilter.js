/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import React, {useState, useEffect} from 'react';
import classnames from 'classnames';
import {parseISO, startOfDay, endOfDay} from 'date-fns';
import {Button, MenuItem, MenuItemSelectable, MenuItemDivider} from '@carbon/react';
import {Filter} from '@carbon/icons-react';
import {MenuDropdown} from '@camunda/camunda-optimize-composite-components';

import {format, BACKEND_DATE_FORMAT} from 'dates';
import {Popover, DatePicker} from 'components';
import {t} from 'translation';

import RollingFilter from './RollingFilter';

import './DateFilter.scss';

export default function DateFilter({
  filter,
  setFilter,
  icon: Icon = Filter,
  children,
  emptyText = t('common.select'),
  title,
}) {
  const isFixed = filter?.type === 'fixed' || filter?.type === 'rolling';

  const [showFixedPopover, setShowFixedPopover] = useState(isFixed);
  const [autoOpen, setAutoOpen] = useState(false);
  const startDate = filter?.type === 'fixed' && filter?.start ? parseISO(filter.start) : null;
  const endDate = filter?.type === 'fixed' && filter?.end ? parseISO(filter.end) : null;

  const [fixedType, setFixedType] = useState(getFixedType(startDate, endDate));

  useEffect(() => {
    setShowFixedPopover(isFixed);
    setAutoOpen(false);
  }, [isFixed]);

  useEffect(() => {
    setFixedType(getFixedType(startDate, endDate));
  }, [endDate, filter, startDate]);

  function setRelativeFilter(unit, past) {
    setFilter({
      type: 'relative',
      start: {value: past ? 1 : 0, unit},
      end: null,
      excludeUndefined: false,
      includeUndefined: false,
    });
  }

  function isFilter(unit, past) {
    return filter?.start.value === (past ? 1 : 0) && filter?.start.unit === unit;
  }

  function getFilterName(filter) {
    if (!filter || filter?.type === 'fixed' || filter?.type === 'rolling') {
      return emptyText;
    }

    if (isFilter('days')) {
      return t('common.filter.dateModal.unit.today');
    }

    if (isFilter('days', true)) {
      return t('common.filter.dateModal.unit.yesterday');
    }

    return (
      t(`dashboard.filter.date.${filter.start.value ? 'last' : 'this'}`) +
      ' ' +
      t('dashboard.filter.date.units.' + filter.start.unit)
    );
  }

  function getFixedDateFilterName(filter) {
    if (!filter) {
      if (fixedType === 'rolling') {
        return emptyText;
      }

      return t('common.filter.dateModal.unit.' + fixedType);
    }

    if (fixedType === 'rolling') {
      const value = filter.start.value;
      const unit = filter.start.unit.slice(0, -1);

      let prefix = 'all';
      if (+value > 1) {
        prefix = 'plural';
      } else if (['weeks', 'minutes', 'hours'].includes(unit)) {
        prefix = 'week';
      }
      return (
        <>
          {t('common.filter.list.operators.occurs')}{' '}
          {t('common.filter.dateModal.preview.last.' + prefix)}{' '}
          {`${+value} ${t(`common.unit.${unit}.${+value === 1 ? 'label' : 'label-plural'}`)}`}
        </>
      );
    }

    return (
      <>
        {t('common.filter.dateModal.unit.' + fixedType)}{' '}
        {fixedType !== 'before' && startDate && format(startDate, 'yyyy-MM-dd')}
        {fixedType === 'between' && <span className="to">{t('common.filter.dateModal.to')}</span>}
        {fixedType !== 'after' && endDate && format(endDate, 'yyyy-MM-dd')}
      </>
    );
  }

  function openDatePicker(type) {
    setFilter();
    setFixedType(type);
    setShowFixedPopover(true);
    setAutoOpen(true);
  }

  function openRollingFilter() {
    setFilter({
      type: 'rolling',
      start: {value: 2, unit: 'days'},
      end: null,
      excludeUndefined: false,
      includeUndefined: false,
    });
    setFixedType('rolling');
    setShowFixedPopover(true);
    setAutoOpen(true);
  }

  function updateRolling({value, unit}) {
    setFilter({
      type: 'rolling',
      start: {value: value || filter?.start?.value, unit: unit || filter?.start?.unit},
      end: null,
      excludeUndefined: false,
      includeUndefined: false,
    });
  }

  return (
    <div className={classnames('DateFilter__Dashboard', {fixed: filter?.type === 'fixed'})}>
      <div className="title">
        {title}
        {children}
      </div>
      {showFixedPopover && filter?.type !== 'relative' ? (
        <Popover
          isTabTip
          trigger={
            <Popover.ListBox size="sm">
              <Icon className={classnames('indicator', {active: filter})} />
              {getFixedDateFilterName(filter)}
            </Popover.ListBox>
          }
          autoOpen={autoOpen}
        >
          {fixedType === 'rolling' ? (
            <RollingFilter filter={filter} onChange={updateRolling} />
          ) : (
            <DatePicker
              type={fixedType}
              forceOpen
              initialDates={{startDate, endDate}}
              onDateChange={({startDate, endDate, valid}) => {
                if (valid) {
                  setFilter({
                    type: 'fixed',
                    start: startDate ? format(startOfDay(startDate), BACKEND_DATE_FORMAT) : null,
                    end: endDate ? format(endOfDay(endDate), BACKEND_DATE_FORMAT) : null,
                    excludeUndefined: false,
                    includeUndefined: false,
                  });
                }
              }}
            />
          )}
          <hr />
          <Button
            size="sm"
            kind="ghost"
            className="reset-button"
            onClick={() => {
              setFilter();
              setShowFixedPopover(false);
            }}
          >
            {t('common.off')}
          </Button>
        </Popover>
      ) : (
        <MenuDropdown
          size="sm"
          label={
            <>
              <Icon className={classnames('indicator', {active: filter})} />
              {getFilterName(filter)}
            </>
          }
          className="filterMenu"
          menuTarget={document.querySelector('.fullscreen')}
        >
          <MenuItem
            label={t('common.filter.dateModal.unit.between')}
            onClick={() => openDatePicker('between')}
          />
          <MenuItem
            label={t('common.filter.dateModal.unit.after')}
            onClick={() => openDatePicker('after')}
          />
          <MenuItem
            label={t('common.filter.dateModal.unit.before')}
            onClick={() => openDatePicker('before')}
          />
          <MenuItemSelectable
            label={t('common.filter.dateModal.unit.today')}
            selected={isFilter('days')}
            onChange={() => setRelativeFilter('days')}
          />
          <MenuItemSelectable
            label={t('common.filter.dateModal.unit.yesterday')}
            selected={isFilter('days', true)}
            onChange={() => setRelativeFilter('days', true)}
          />
          <MenuItemSelectable
            label={t('dashboard.filter.date.last')}
            selected={filter?.start.value === 1 && !isFilter('days', true)}
          >
            {['weeks', 'months', 'years', 'quarters'].map((unit) => (
              <MenuItemSelectable
                key={unit}
                label={t('dashboard.filter.date.units.' + unit)}
                selected={isFilter(unit, true)}
                onChange={() => setRelativeFilter(unit, true)}
              />
            ))}
          </MenuItemSelectable>
          <MenuItemSelectable
            label={t('dashboard.filter.date.this')}
            selected={filter?.start.value === 0 && !isFilter('days')}
          >
            {['weeks', 'months', 'years', 'quarters'].map((unit) => (
              <MenuItemSelectable
                key={unit}
                label={t('dashboard.filter.date.units.' + unit)}
                selected={isFilter(unit)}
                onChange={() => setRelativeFilter(unit)}
              />
            ))}
          </MenuItemSelectable>
          <MenuItem
            label={t('common.filter.dateModal.unit.custom')}
            onClick={() => openRollingFilter()}
          />
          <MenuItemDivider />
          <MenuItem disabled={!filter} label={t('common.off')} onClick={() => setFilter()} />
        </MenuDropdown>
      )}
    </div>
  );
}

const getFixedType = (start, end) => {
  if (start && end) {
    return 'between';
  } else if (start) {
    return 'after';
  } else if (end) {
    return 'before';
  } else {
    return 'rolling';
  }
};
