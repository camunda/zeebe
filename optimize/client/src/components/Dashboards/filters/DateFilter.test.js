/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import React, {runAllEffects} from 'react';
import {shallow} from 'enzyme';
import {MenuItem, MenuItemSelectable} from '@carbon/react';
import {MenuDropdown} from '@camunda/camunda-optimize-composite-components';

import {DatePicker} from 'components';

import DateFilter from './DateFilter';
import RollingFilter from './RollingFilter';

const props = {
  filter: null,
  setFilter: jest.fn(),
  type: 'instanceStartDate',
};

const todayFilter = {
  type: 'relative',
  start: {value: 0, unit: 'days'},
  end: null,
  includeUndefined: false,
  excludeUndefined: false,
};

beforeEach(() => {
  props.setFilter.mockClear();
});

it('should contain a dropdown to set startDateFilter', () => {
  const node = shallow(<DateFilter {...props} />);

  expect(node.find(MenuDropdown)).toExist();

  node.find(MenuItemSelectable).at(0).simulate('change');

  expect(props.setFilter).toHaveBeenCalledWith(todayFilter);
});

it('should show a datepicker when switching to fixed state', () => {
  const node = shallow(<DateFilter {...props} />);

  node.find(MenuItem).at(0).simulate('click');

  expect(node.find(DatePicker)).toExist();
});

it('should invoke setFilter when updating the rolling filter', () => {
  const spy = jest.fn();
  const filter = {
    type: 'rolling',
    start: {value: 2, unit: 'days'},
    end: null,
    excludeUndefined: false,
    includeUndefined: false,
  };
  const node = shallow(<DateFilter {...props} setFilter={spy} />);

  node.setProps({filter});

  const options = node.find(MenuItem);
  options.at(options.length - 2).simulate('click');
  expect(spy).toHaveBeenCalledWith(filter);

  node.find(RollingFilter).simulate('change', {value: '5', unit: 'months'});
  expect(spy).toHaveBeenCalledWith({...filter, start: {value: '5', unit: 'months'}});
});

it('should show the dropdown again after an external reset', () => {
  const node = shallow(<DateFilter {...props} />);

  node.find(MenuItem).at(0).simulate('click');
  node.setProps({resetTrigger: true});

  runAllEffects();

  expect(node.find(DatePicker)).not.toExist();
  expect(node.find(MenuDropdown)).toExist();
});

it('should show the filter state', () => {
  const node = shallow(<DateFilter {...props} filter={todayFilter} />);

  expect(node.find(MenuDropdown).prop('label')).toMatchSnapshot();
  expect(node.find(MenuItemSelectable).at(0)).toHaveProp('selected', true);
});

it('should reset the filter state', () => {
  const node = shallow(<DateFilter {...props} filter={todayFilter} />);

  node.find(MenuItem).last().simulate('click');

  expect(props.setFilter).toHaveBeenCalledWith();
});

it('should disable the reset button if no filter is set', () => {
  const node = shallow(<DateFilter {...props} />);

  expect(node.find(MenuItem).last()).toHaveProp('disabled', true);
});

it('should allow providing a custom icon and empty text', () => {
  const node = shallow(<DateFilter {...props} icon="customIcon" emptyText="customText" />);

  expect(node.find(MenuDropdown).prop('label')).toMatchSnapshot();
});

it('should render children', () => {
  const node = shallow(
    <DateFilter {...props}>
      <div className="childContent" />
    </DateFilter>
  );

  expect(node.find('.childContent')).toExist();
});
