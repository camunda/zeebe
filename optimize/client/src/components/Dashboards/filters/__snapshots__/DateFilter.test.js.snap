// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should allow providing a custom icon and empty text 1`] = `
<React.Fragment>
  <customIcon
    className="indicator"
  />
  customText
</React.Fragment>
`;

exports[`should show the filter state 1`] = `
<React.Fragment>
  <ForwardRef(Filter)
    className="indicator active"
  />
  Today
</React.Fragment>
`;
